Source/Dafny/Plugin.cs

```diff
@@ -0,0 +1,82 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using Microsoft.Dafny.Plugins;
+
+namespace Microsoft.Dafny;
+
+/// <summary>
+/// This class is internal to Dafny. It wraps an assembly and an extracted configuration from this assembly,
+/// The configuration provides the methods to parse command-line arguments and obtain Rewriters 
+/// </summary>
+public class Plugin {
+  public Configuration Configuration;
+
+  private readonly string path;
+  private readonly Assembly assembly;
+  private System.Type[] rewriterTypes;
+
+  class AutomaticConfiguration : Configuration {
```

Comment:
Consider using a `record` instead of a class. Why is this a nested class?

Reply:
> Nothing else than the class Plugin needs this class which is very short, so it makes sense to keep it as a nested class.

Okay. In my experience working on C# codebases, nested classes are only used when access to private members of the containing class is needed, but that's not the case here. For a class that's used by another class through public members, I'm used to using top-level classes, which compared to nested classes may separate concerns more and reduce indentation. I don't have any hard arguments for this way of using nested classes, so feel free to keep it like it is, but consider this an extra data point ^,^

> Based on your suggestion, I considered using a record instead of a class, but as it is extending an abstract class, as far as I know, this is not possible.

You'd have to change the `abstract class` into an `abstract record`

Reply:
I'm going to revert the use of "abstract record" for PluginConfiguration back to an "abstract class", after testing its usability.
Since the goal is for developers to extend the class PluginConfiguration, if I mark it as an abstract record, they will be forced to implement internal methods such as `PluginConfiguration.<Clone>$()`, and the other goodies that record routinely offer, that abstract record actually require. It would be fine if they developed their plugins in c#, but the dotnet ecosystem has many other languages which do not know about records (e.g. F#).

Reply:
Makes sense, nice UX research :)